src/import/load-costume.js

```diff
@@ -32,6 +32,53 @@ const loadVector_ = function (costume, runtime, rotationCenter, optVersion) {
     });
 };
 
+const canvasPool = (function () {
+    /**
+     * A pool of canvas objects that can be reused to reduce memory
+     * allocations. And time spent in those allocations and the later garbage
+     * collection.
+     */
+    class CanvasPool {
+        constructor () {
+            this.pool = [];
+            this.clearSoon = null;
+        }
+
+        /**
+         * After a short wait period clear the pool to let the VM collect
+         * garbage.
+         */
+        clear () {
+            if (!this.clearSoon) {
+                this.clearSoon = new Promise(resolve => setTimeout(resolve, 1000))
+                    .then(() => {
+                        this.pool.length = 0;
+                        this.clearSoon = null;
+                    });
+            }
+        }
+
+        /**
+         * Return a canvas. Create the canvas if the pool is empty.
+         * @returns {HTMLCanvasElement} A canvas element.
+         */
+        create () {
+            return this.pool.pop() || document.createElement('canvas');
+        }
+
+        /**
+         * Release the canvas to be reused.
+         * @param {HTMLCanvasElement} A canvas element.
+         */
+        release (canvas) {
+            this.clear();
```

Comment:
Do you mean to call `clear()` here? This means that the canvases that enter the pool will last at most one JS tick. I guess that would help with thrashing during one tick but it seems like letting them last for several ticks would be even better...? Or is the idea that you're just trying to avoid one JS tick creating many canvas elements that will all be eligible for GC by the end of the tick?

Reply:
I agree holding canvases for several ticks may be better for mixing loading performance and memory used as a simple strategy.

I decided to simplify this Pool class where I could since I don't intend this specific implementation to be used outside of this file. Really I only wanted this class to handle the general improvement of reusing canvases and some safety if I'm wrong about one fact of this file.

As is, load-costume will only have 0 or 1 canvases in this pool. As the promise chain is written it will "create", use, and release the canvas synchronously every time without leaving the javascript context. That is to say the promise chain all resolves after the load event of the images and browsers will exhaust those resolve jobs before returning to the native context to possibly emit another load event. (Of course this could change from a change to the file or other change, or if I'm just wrong. In those cases we have the safety of the list of canvases hear.)

-----

I haven't spent enough time thinking on architecture to better solve this. It's clear to me that we need a large behaviour managing this. But there is a bunch of needs. I think I want to propose two things a TaskLoop and ReferenceManager on VM or something to manage large processing jobs or large memory resources.

Reply:
*Facepalm*

Crud pretty sure what I just wrote out means this version of the canvas pool stand alone doesn't help. I've done testing on a bunch of changes standalone and others grouped with other changes and I can't keep track of what things depend on each other anymore clearly.

...

Oh wait. No thats not right. I already thought of this.

https://github.com/LLK/scratch-vm/pull/1970/files#diff-a63caa2e62f9dd4327204d2dd61b4b3fR53

The pool is cleaned up after 1 second after a canvas is released when no work to clear the pool is queued.